File: 1335-maximum-candies-allocated-to-k-children/maximum-candies-allocated-to-k-children.java
class Solution {
    private boolean canDistribute(int[]candies,long k,long mid)
    {
        long cnt=0;
        for(int i=0;i<candies.length;i++)
        {
            long x=candies[i]/mid;
            cnt+=x;
            if(cnt>=k) return true;
    
        }

        return false;
    }
    public int maximumCandies(int[] candies, long k) 
    {
        if(candies.length==0) return 0;
        long low=1;
        long high=-1;
        for(int x:candies)
        {
            if(x>high) high=x;
        }

        while(low<=high)
        {
            long mid=low+(high-low)/2;
            if(canDistribute(candies,k,mid))
            {
                low=mid+1;
            }
            else
            {
                high=mid-1;
            }
        }

        return (int)high;
        
    }
}

File: 1335-maximum-candies-allocated-to-k-children/maximum-candies-allocated-to-k-children.cpp
class Solution {
// private:
// {
//     bool isPossible(vector<int>& candies,long long k,long long candy)
//     {
//         int n=candies.size();
//         long long child=0;
//         for(int i=0;i<n;i++)
//         {
//             child+=candies[i]/candy;
//             if(child>=k) return true;
//         }

//         return false;
//     }

// };
public:
    bool isPossible(vector<int>& candies,long long k,long long candy)
    {
        int n=candies.size();
        long long child=0;
        for(int i=0;i<n;i++)
        {
            child+=candies[i]/candy;
            if(child>=k) return true;
        }

        return false;
    }

    int maximumCandies(vector<int>& candies, long long k) 
    {
        long long low=1;
        long long totalSum=0;
        for(int x:candies)
        {
            totalSum+=x;
        }
        long long high=totalSum/k;
        if(totalSum<k) return 0;
        int ans=0;
        while(low<=high)
        {
            long long mid=low+(high-low)/2;
            if(isPossible(candies,k,mid))
            {
                ans=(int)mid;
                low=mid+1;
            }
            else
            {
                high=mid-1;
            }

        }
        return ans;

        
    }
};